File: src/main/java/com/jcloisterzone/feature/modifier/FeaturePointerAddModifier.java
package com.jcloisterzone.feature.modifier;

import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.ModifiedFeature;
import com.jcloisterzone.game.setup.SetupQuery;
import io.vavr.collection.Stream;

public class FeaturePointerAddModifier<F extends Feature> extends FeatureModifier<Stream<FeaturePointer>> {

	private final Class<F> featureClass;
	
    public FeaturePointerAddModifier(String selector, SetupQuery enabledBy, Class<F> featureClass) {
        super(selector, enabledBy);
        this.featureClass = featureClass;
    }

    @Override
    public Stream<FeaturePointer> mergeValues(Stream<FeaturePointer> a, Stream<FeaturePointer> b) {
    	if (a == null) {
    		return b;
    	}
    	if (b == null) {
    		return a;
    	}
        return a.appendAll(b);
    }

    @Override
    public Stream<FeaturePointer> valueOf(String attr) {
    	String[] locations = attr.trim().split("\\s+");

    	return Stream.of(locations).map(s -> new FeaturePointer(Position.ZERO, this.featureClass, Location.valueOf(s)));
    }
}

File: src/main/ts/com/jcloisterzone/feature/modifier/FeaturePointerAddModifier.ts
import { Stream } from "../../../../io/vavr/SeqTypes.js";
import { type ClassToken } from "../../../../lang/Class.js";
import { Location } from "../../board/Location.js";
import { Position } from "../../board/Position.js";
import { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import type { Feature } from "../Feature.js";
import type { SetupQuery } from "../../game/setup/SetupQuery.js";
import { FeatureModifier } from "./FeatureModifier.js";

/** Modifier accumulating feature pointers (e.g. gamblers-luck shields). */
export class FeaturePointerAddModifier<F extends Feature> extends FeatureModifier<
  Stream<FeaturePointer>
> {
  private readonly featureClass: ClassToken<F>;

  constructor(selector: string, enabledBy: SetupQuery | null, featureClass: ClassToken<F>) {
    super(selector, enabledBy);
    this.featureClass = featureClass;
  }

  mergeValues(a: Stream<FeaturePointer>, b: Stream<FeaturePointer>): Stream<FeaturePointer> {
    if (a === null) return b;
    if (b === null) return a;
    return a.appendAll(b) as Stream<FeaturePointer>;
  }

  valueOf(attr: string): Stream<FeaturePointer> {
    const locations = attr.trim().split(/\s+/);
    return Stream.ofAll(
      locations.map((s) => new FeaturePointer(Position.ZERO, this.featureClass, Location.valueOf(s))),
    );
  }
}
